src/resmom/mom_pmix.c

```diff
@@ -0,0 +1,1537 @@
+		pthread_cond_init(&(l)->cond, NULL); \
+		(l)->active = true; \
+		(l)->status = PMIX_SUCCESS; \
+	} while(0)
+
+#define PBS_PMIX_DESTRUCT_LOCK(l) \
+	do { \
+		pthread_mutex_destroy(&(l)->mutex); \
+		pthread_cond_destroy(&(l)->cond); \
+	} while(0)
+
+#define PBS_PMIX_WAIT_THREAD(lck) \
+	do { \
+		pthread_mutex_lock(&(lck)->mutex); \
+		while ((lck)->active) { \
+			pthread_cond_wait(&(lck)->cond, &(lck)->mutex); \
+		} \
+		pthread_mutex_unlock(&(lck)->mutex); \
+	} while(0)
+
+#define PBS_PMIX_WAKEUP_THREAD(lck) \
+	do { \
+		pthread_mutex_lock(&(lck)->mutex); \
+		(lck)->active = false; \
+		pthread_cond_broadcast(&(lck)->cond); \
+		pthread_mutex_unlock(&(lck)->mutex); \
+	} while(0)
+
+/**
+ * @brief
+ * This callback function is invoked by the PMIx library after it
+ * has been notified a process has exited.
+ *
+ * @param[in] status - exit status of the process
+ * @param[in] cbdata - opaque callback data passed to the caller
+ *
+ * @retval void
+ *
+ * @note
+ * This function may be superfluous, in which case the call to
+ * PMIx_Notify_event() should be passed NULL in pbs_pmix_notify_exit()
+ * for its callback and this funtion removed. It has been left in for
+ * the time being so that the log shows it being called.
+ */
+static void
+pbs_pmix_notify_exit_cb(pmix_status_t status, void *cbdata)
+{
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "called");
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "returning");
+}
+
+/**
+ * @brief
+ * Notify PMIx that a task has exited by constructing a PMIx info
+ * array and passing it to PMIx_Notify_event.
+ *
+ * @param[in] pjob - pointer to the job structure for the exited task
+ * @param[in] exitstat - numeric exit status of the task
+ * @param[in] msg - optional message supplied by the caller
+ *
+ * @retval void
+ */
+void
+pbs_pmix_notify_exit(job *pjob, int exitstat, char *msg)
+{
+	pmix_status_t status;
+	pmix_info_t *pinfo;
+	size_t ninfo;
+	pmix_proc_t procname, procsrc;
+	bool flag = true;
+
+	log_event(PBSEVENT_DEBUG3, 0, LOG_DEBUG, __func__, "called");
+	if (!pjob) {
+		log_event(PBSEVENT_DEBUG, 0, LOG_ERR, __func__,
+			"No job supplied, returning");
+		return;
+	}
+	snprintf(log_buffer, sizeof(log_buffer),
+		"Setting up the info array for termination");
+	log_event(PBSEVENT_JOB, PBS_EVENTCLASS_JOB, LOG_INFO,
+		pjob->ji_qs.ji_jobid, log_buffer);
+	ninfo = 3;
+	PMIX_INFO_CREATE(pinfo, ninfo);
+	/* Ensure this only goes to the job terminated event handler */
+	PMIX_INFO_LOAD(&pinfo[0], PMIX_EVENT_NON_DEFAULT, &flag, PMIX_BOOL);
+	/* Provide the exit status of the application */
+	PMIX_INFO_LOAD(&pinfo[1], PMIX_JOB_TERM_STATUS, &exitstat, PMIX_STATUS);
+	/* Provide the rank */
+	PMIX_LOAD_PROCID(&procname, pjob->ji_qs.ji_jobid, PMIX_RANK_WILDCARD);
+	PMIX_INFO_LOAD(&pinfo[2], PMIX_EVENT_AFFECTED_PROC, &procname, PMIX_PROC);
+	/* Provide the message if provided */
+	if (msg) {
+		ninfo++;
+		PMIX_INFO_LOAD(&pinfo[3], PMIX_EVENT_TEXT_MESSAGE, msg, PMIX_STRING);
```

Comment:
we allocated only 3 pmix_info_t through PMIX_INFO_CREATE and we are using the 4th element here.

Reply:
could you please reply to or address this comment.

Reply:
I believe it has been addressed. The info array will contain four entries if a message is present.

Reply:
hey @mike0042 - not sure why I did not see the updated code.

Reply:
I thought that might be the case. No problem, and thank you for the review.